ui/statistics_button.js

```diff
@@ -0,0 +1,250 @@
+goog.require('shaka.ui.Locales');
+goog.require('shaka.ui.Localization');
+goog.require('shaka.ui.Utils');
+goog.require('shaka.util.Dom');
+goog.require('shaka.util.Timer');
+goog.requireType('shaka.ui.Controls');
+
+
+/**
+ * @extends {shaka.ui.Element}
+ * @final
+ * @export
+ */
+shaka.ui.StatisticsButton = class extends shaka.ui.Element {
+  /**
+   * @param {!HTMLElement} parent
+   * @param {!shaka.ui.Controls} controls
+   */
+  constructor(parent, controls) {
+    super(parent, controls);
+
+    /** @private {!HTMLButtonElement} */
+    this.button_ = shaka.util.Dom.createButton();
+    this.button_.classList.add('shaka-statistics-button');
+
+    /** @private {!HTMLElement} */
+    this.icon_ = shaka.util.Dom.createHTMLElement('i');
+    this.icon_.classList.add('material-icons-round');
+    this.icon_.textContent =
+      shaka.ui.Enums.MaterialDesignIcons.STATISTICS_ON;
+    this.button_.appendChild(this.icon_);
+
+    const label = shaka.util.Dom.createHTMLElement('label');
+    label.classList.add('shaka-overflow-button-label');
+
+    /** @private {!HTMLElement} */
+    this.nameSpan_ = shaka.util.Dom.createHTMLElement('span');
+    label.appendChild(this.nameSpan_);
+
+    /** @private {!HTMLElement} */
+    this.stateSpan_ = shaka.util.Dom.createHTMLElement('span');
+    label.appendChild(this.stateSpan_);
+
+    this.button_.appendChild(label);
+
+    this.parent.appendChild(this.button_);
+
+    /** @private {!HTMLElement} */
+    this.container_ = shaka.util.Dom.createHTMLElement('div');
+    this.container_.classList.add('shaka-no-propagation');
+    this.container_.classList.add('shaka-show-controls-on-mouse-over');
+    this.container_.classList.add('shaka-statistics-container');
+    this.container_.classList.add('shaka-hidden');
+
+    const controlsContainer = this.controls.getControlsContainer();
+    controlsContainer.appendChild(this.container_);
+
+    /** @private {!Array} */
+    this.statisticsList_ = [];
+
+    /** @private {!Array} */
+    this.skippedStats_ = ['stateHistory', 'switchHistory'];
+
+    /** @private {!Object.<string, number>} */
+    this.currentStats_ = this.player.getStats();
+
+    /** @private {!Object.<string, HTMLElement>} */
+    this.displayedElements_ = {};
+
+    /** @private {!Object.<string, string>} */
+    this.parseFrom_ = {
+      'width': 'px',
+      'height': 'px',
+      'completionPercent': 'percent',
+      'bufferingTime': 'seconds',
+      'drmTimeSeconds': 'seconds',
+      'licenseTime': 'seconds',
+      'liveLatency': 'seconds',
+      'loadLatency': 'seconds',
+      'manifestTimeSeconds': 'seconds',
+      'estimatedBandwidth': 'bits',
+      'streamBandwidth': 'bits',
+      'maxSegmentDuration': 'time',
+      'pauseTime': 'time',
+      'playTime': 'time',
+      'corruptedFrames': 'frames',
+      'decodedFrames': 'frames',
+      'droppedFrames': 'frames',
+    };
+
+    /** @private {!Object.<string, Function>} */
+    this.parseTo_ = {
```

Comment:
There are a lot of potential points of failure here; everything in `parseFrom_` and `parseTo_` is a text string, so if we misspelled the name of a key (e.g. `playtime` instead of `playTime`), or the name of an intermediate value (e.g. `farmes` instead of `frames`), it wouldn't be picked up by the compiler. That could be relevant if we ever add more things to the statistics object.

I'd say:

1. Instead of having the `parseTo_` method, just make local variables with the conversion functions, and pass them in to the `parseFrom_` object. So it'd be like: `this.parseFrom_ = {'width': pixelConverter, 'height': pixelConverter, 'percent': percentConverter, ... }`
2. Make a new test that tests every single key in the statistics object, rather than just the four you have right now. It could go over all of the keys of a statistics object using `for key in`, so as to automatically update if we add new statistics. It wouldn't need to see if they have sane values, like your current test, it could just check if the elements exist and have text of some sort in them.

Reply:
Thank you Theodab, I implemented both improvements on my last commit!